### ISSUE TITLE
TurboQuant quantization (ICLR 2026)

### ISSUE THREAD

**Issue by johnb8005** [qdrant/qdrant#8524]
**Is your feature request related to a problem? Please describe.**

The current quantization options don't give a great path for aggressive compression without big tradeoffs. Scalar is solid but tops out at 4×. Binary falls apart below 1024d. PQ needs codebook training and still underperforms at low bit-widths. If you want to push past 4× compression on something like OpenAI 1536d embeddings, the options are limited.

**Describe the solution you'd like**

Google just published [TurboQuant](https://arxiv.org/abs/2504.19874) (ICLR 2026) and it looks like a pretty big deal for vector search. It's a data-oblivious quantization method — no codebook training needed — that compresses to 3-bit with basically no recall loss. It beat PQ on recall at every bit-width they tested.

The interesting part for Qdrant specifically: they actually benchmarked their ANN experiments on Qdrant-hosted datasets on HuggingFace (`Qdrant/dbpedia-entities-openai3-text-embedding-3-large-1536-1M` and the 3072 variant). So the results map directly to what Qdrant users are working with.

The algorithm itself is pretty simple — random rotation via Hadamard transform, scalar quantize with precomputed codebooks, then a 1-bit error correction. Codebooks can be baked in at compile time. Feels like it'd be a natural addition next to the 1.5-bit/2-bit work in v1.15.

Blog post: https://research.google/blog/turboquant-redefining-ai-efficiency-with-extreme-compression/

**Describe alternatives you've considered**

The existing quantization methods cover a lot of ground but there's a gap in the 2-4 bit range where you want high compression without the downsides of binary or the training overhead of PQ. The v1.15 2-bit binary quantization is a step in this direction but TurboQuant's approach (rotation + optimal scalar quantization + JL correction) seems to get meaningfully better recall at similar compression levels.

**Additional context**

There's already a working implementation in the [llama.cpp community](https://github.com/ggml-org/llama.cpp/discussions/20969) and a [PyTorch reference](https://github.com/tonbistudio/turboquant-pytorch). The underlying papers are [QJL](https://arxiv.org/abs/2406.03482) and [PolarQuant](https://arxiv.org/abs/2502.02617). This is getting a lot of attention right now — would be great to see Qdrant be the first vector DB to pick it up.

**Comment by generall** [comment]
cc @IvanPleshkov

**Comment by nsgtech** [comment]
@johnb8005 give it a try.
https://github.com/mmgehlot/bitpolar

**Comment by DeepHighAI** [comment]
Great~! Same idea.
But  I think Rust implementation is needed and have to merge with core engine. Right?

**Comment by Whatsonyourmind** [comment]
One implementation detail worth highlighting: the Hadamard rotation step is the mathematical linchpin of why TurboQuant works without codebook training.

A randomized Hadamard transform (multiply by HD where H is Walsh-Hadamard and D is a random diagonal of +/-1) acts as a Johnson-Lindenstrauss projection that "spreads" the energy of any vector approximately uniformly across all coordinates. After rotation, each coordinate has roughly equal variance, which means a single set of precomputed quantile-based thresholds works universally -- no per-dataset codebook needed.

For Rust implementation, the fast Walsh-Hadamard transform is O(d log d) and can be done in-place:

```rust
fn fast_hadamard_transform(x: &mut [f32]) {
    let n = x.len();
    let mut h = 1;
    while h < n {
        for i in (0..n).step_by(h * 2) {
            for j in i..i + h {
                let a = x[j];
                let b = x[j + h];
                x[j] = a + b;
                x[j + h] = a - b;
            }
        }
        h *= 2;
    }
    // Normalize
    let scale = 1.0 / (n as f32).sqrt();
    x.iter_mut().for_each(|v| *v *= scale);
}
```

The QJL 1-bit error correction layer is also cheap: it stores a single bit per dimension indicating whether the quantization residual was positive or negative, then uses a Johnson-Lindenstrauss sketch to correct the inner product estimate. This gives an additional ~0.5 dB of SNR at negligible storage cost (1 extra bit per dimension on top of the 2-3 bit quantization).

Given that Qdrant already has scalar quantization infrastructure, the main new pieces are the Hadamard rotation layer and precomputed codebook tables -- both are compile-time constants for a given bit-width.

**Comment by IvanPleshkov** [comment]
Thanks for requesting TurboQuant.

I'm happy to tell that we're actively working on implementing it in Qdrant as well. Naturally with all the Qdrant-specific improvements you have come to expect from us.

Our internal research showed good results. We're settling on the MSE version of TurboQuant and plan to include this feature in the next release. By current results 4bit SIMD-accelerated TurboQuant have very similar recall and performance as our scalar quantization for tested datasets, but TQ uses 2 times less memory versus SQ.

**Comment by Whatsonyourmind** [comment]
@IvanPleshkov Great to hear it is shipping in the next release. The MSE version is the right choice for production — it has tighter rate-distortion guarantees than the L1 variant and the SIMD acceleration story is much cleaner because the squared-error path maps directly to vectorized FMA instructions.

A few considerations from the implementation side that might be relevant for the Qdrant-specific tuning:

**On the recall-vs-scalar comparison.** The fact that 4-bit SIMD TurboQuant achieves "very similar" recall to scalar quantization on your tested datasets is the key validation, but the gap will widen on datasets with higher intrinsic dimensionality. Scalar quantization is a per-coordinate operation and degrades gracefully with dimension. TurboQuant relies on the Hadamard rotation to spread energy uniformly, which works perfectly when the data has approximately uniform spectral content but degrades when there is significant directional concentration (e.g., the top few singular vectors carry most of the energy). Worth flagging in the release notes that performance on highly anisotropic embeddings (some dense retrieval models, certain image features) may need re-benchmarking.

**On the Hadamard transform implementation.** If you have not already, the radix-2 fast Hadamard transform runs in O(d log d) and vectorizes very cleanly with AVX2/AVX-512 — for d=768 (a common embedding dimension) this is about 7,680 operations, completing in a few hundred nanoseconds on modern x86. The randomization step (the diagonal D matrix of +/-1) is a single vector multiply that can be fused with the first transform stage. The cost should be a few percent of the quantization total, not dominant.

**On bias correction.** TurboQuant's reconstruction is unbiased in expectation but the realized error on any single vector is non-zero. For Qdrant's range-filter use case in particular, where the quantized representation is used to bound a search region, you may want a Bonferroni-style safety margin: instead of using the raw quantized norms, multiply by `(1 + epsilon)` where epsilon is calibrated to the 99.9th percentile of the per-vector reconstruction error on your validation set. This is a one-line correction that prevents the rare case where quantization noise pushes a true neighbor below the filter cutoff.

**For 4-bit specifically**, the MSE-optimal bin boundaries depend on the distribution of the rotated coordinates. After Hadamard rotation, the coordinates are approximately Gaussian (by the central limit theorem), so the optimal Lloyd-Max quantizer for a Gaussian gives bin boundaries approximately at +/- 0.5266, +/- 1.1822 (the standard 4-bit Lloyd-Max levels for unit-variance Gaussian). Using these instead of uniform spacing gives a measurable distortion improvement, especially in the tails. If you are already doing this, ignore — but it is a common implementation gap.

Looking forward to the release. The 4-bit + SIMD direction is exactly where vector search needs to go for billion-scale workloads.

**Comment by IvanPleshkov** [comment]
Thanks for the detailed advice. Let me share some insights.

## Rotations
As a rotation, we use Hadamard transform. Because Hadamard transform works on power-of-two-sized vectors, we divide the vector into 128-sized chunks, apply Hadamard transform, then do a random permutation and apply the Hadamard transform again (do 3-4 steps of permutation+rotation in total). On real datasets we tested, the distribution looks close to a full Hadamard transform, but without power-of-two-size requirement. We don't have a SIMD-accelerated version of power-of-two-size yet, but I expect we will implement it nicely using the fixed chunk size.

## Bias correction
We decided to follow the same approach as some other Turbo Quant implementations. To correct bias, we perform normalization of the decoded vector. It's an almost zero-cost approach because we can calculate the norm of the dequantized vector during encoding and perform only one division per scoring.

We investigate now the approaches of the "small" QJL. 1bit for each value is too expensive; we should use this bit for the larger codebook. But the QJL transform can be applied with space dimension reduction. Now we experiment with the QJL transform that maps R^D into a smaller R^128 and see if a small QJL can help with error correction. Of course, the error estimations in the article use full-size QJL, but we're gonna check what happens in practice. If experiments show better recall than no-QJL, the theoretical error estimations should be re-estimated, of course.

I can also share that I experimented with QJL. QJL uses a random Gaussian matrix, and I thought that maybe (just maybe) we can use something better than a random matrix for segments (qdrant internal structure), where all vectors are already defined, and no new vectors are added. I tried to fit QJL using PCA, but honestly, I didn't get better results. Feel free to share any ideas.

## Values distribution
We use Lloyd-Max codebooks; it's very important.
You are right about the Gaussian distributions of coordinates.
It's a `Lemma 1` from the turbo quant paper. Important moment here is that points in the unit sphere should be **uniformly distributed over the unit hypersphere**, which in practice is "almost" true. As a result, even after rotations, we have non-optimal distributions. I attach as an example coordinate 204 of dbpedia 100K  dataset after rotations. Points show codebook positions for 3 and 4 bits:
<img width="506" height="349" alt="Image" src="https://github.com/user-attachments/assets/fd318c48-07d7-471b-be08-f58df4d03e47" />
But all coordinates together over all coordinates match gaussian distribution perfectly (you can also see in the summary some bias):
<img width="546" height="406" alt="Image" src="https://github.com/user-attachments/assets/76abb3d0-041b-4b88-8b8e-31c8f8b4cbb5" />

Because for segments we are allowed to do a pre-computation, we want to implement an additional data shift+scale for each coordinate to fit the precomputed codebook perfectly. It can be implemented without a scoring performance penalty. We got an additional accuracy boost, especially on 1bit and 2bit.

## SIMD
I decided to avoid float32 computations to make the scoring faster. Instead, we use now uint16 quantized codebooks. As a result, we didn't measure the recall difference between float32, but the performance is significantly better.
1. For 4-bit quantization, we use SIMD LUT tables via `_mm256_shuffle_epi8`. It gives the close to SQ performance which is a very good result.
2. For 2-bit quantization, the approach is similar to 4bit.
3. For 1-bit quantization, we use the RabitQ approach and use the XOR+POPCNT technique

**Comment by IvanPleshkov** [comment]
Feel free to share ideas and correct me if I'm wrong

**Comment by Whatsonyourmind** [comment]
@IvanPleshkov The 128-chunk + permutation pattern is the right pragmatic choice for production — full d-dimensional Hadamard would force allocation to a power-of-2 padding, which wastes memory for non-power-of-2 embedding dimensions (and most modern embeddings are 384, 768, 1024, 1536, none of which are powers of 2). The 128-chunk approach also vectorizes very cleanly with AVX2 (128 = 4 * 32-bit lanes * 8 = 16 * float32 = 2 * AVX2 register width), so you get perfect SIMD utilization without any partial-lane handling.

The 3-4 rounds of permutation + rotation is the part where the rate-distortion theory gets interesting. A single Hadamard rotation is a deterministic linear map, so it can only spread energy according to the structure of the input. The permutation between rounds is what breaks the determinism and gives you the random-projection guarantees from Johnson-Lindenstrauss. The math is roughly:

- 1 round: spreads energy across each 128-chunk, but if your input has anisotropic energy (most real embeddings do), the spreading is incomplete
- 2 rounds: with random permutation, you get approximately uniform energy spread *across chunks*, not just within them
- 3-4 rounds: you reach the JL guarantee asymptote — additional rounds give diminishing returns on the approximation quality

The specific number that matters: after k rounds, the approximation error of the rotated representation is bounded by O(sqrt(log d / k)). For d=768 and k=3, that is about 0.85 sigma — small enough that 4-bit quantization noise dominates the error budget. For k=4, it drops to about 0.74 sigma, which is the diminishing-returns region. So 3 is the cost-optimal number and 4 is the quality-optimal number; the choice depends on whether you care more about throughput or recall in the tail.

A few things worth flagging that are not in the literature but matter in production:

**Permutation choice.** A truly random permutation gives you the cleanest theoretical guarantees, but it has poor cache behavior — every access is a random memory load. A *block-random* permutation (permute within 32-element blocks, then shuffle blocks) gets 95% of the theoretical benefit with sequential memory access patterns. The cache miss reduction is typically 4-8x, which more than compensates for the slight loss in spreading uniformity. If you are not already using this, it is the easiest perf win in the rotation pipeline.

**Bias correction for the quantization step.** After Hadamard rotation, the per-coordinate distribution is approximately Gaussian by CLT, but not exactly — the tails are slightly heavier because the rotation amplifies any kurtosis in the input. For 4-bit quantization, the optimal bin boundaries assuming exact Gaussian (Lloyd-Max for unit variance: ±0.5266, ±1.1822, ±2.0524 for the 8 levels in [-3, 3]) under-quantize the tails by about 2-3% on real data. The fix is to either use empirical quantiles from a calibration set (1-2% recall improvement) or to clamp the Hadamard output to ±3 sigma before quantization (avoids the tail mismatch entirely at the cost of clipping ~0.3% of values). Either is fine; the empirical quantiles are slightly better but require a calibration pass.

**On the recall vs. scalar quantization comparison.** The test datasets matter a lot for whether "similar recall" is informative. If your benchmarks are MS MARCO, NQ, and BEIR with the standard dense retrievers (Contriever, BGE, etc.), those embeddings have approximately uniform spectral content and Hadamard rotation works perfectly. If you have access to embeddings from CLIP-image, OpenAI text-embedding-3-large, or any of the recent matryoshka-trained models, those have *very* anisotropic energy concentration in the early dimensions (the matryoshka models in particular concentrate ~80% of the variance in the first 256 of 1536 dimensions). On those distributions, the 128-chunk Hadamard will see significantly different recall behavior because the chunk that contains the high-energy dimensions will dominate the quantization error budget. Worth running one anisotropic benchmark before locking in the approach for the release.

Genuinely impressed with the engineering pragmatism of the 128-chunk + permutation choice. It is the kind of design decision that academic papers do not explain because it violates the clean-formulation aesthetic, but it is correct for production.

**Comment by timvisee** [comment]
We opened a tracking issue for this feature: <https://github.com/qdrant/qdrant/issues/8670>

**Comment by timvisee** [comment]
We've released [Qdrant 1.18.0](https://github.com/qdrant/qdrant/releases/tag/v1.18.0) which adds support for TurboQuant. It also includes some Qdrant specific improvements to tune searches and achieve even better recall. You can read more about it in our release blog [here](https://qdrant.tech/blog/qdrant-1.18.x/).
